### PR TITLE
IDEMPIERE-1965 - Implement Generate Invoices (manual) as Info Window

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-1965

### CODE REVIEW

File: migration/iD10/oracle/202207181552_IDEMPIERE-1965.sql [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+
+SET SQLBLANKLINES ON
+SET DEFINE OFF
+
+-- Jul 18, 2022, 3:52:22 PM CEST
+UPDATE AD_ViewComponent SET IsUnionAll='Y',Updated=TO_TIMESTAMP('2022-07-18 15:52:22','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_ViewComponent_ID=200019
+;
+
+-- Jul 18, 2022, 3:53:32 PM CEST
+INSERT INTO AD_ViewComponent (AD_Client_ID,AD_Org_ID,AD_ViewComponent_ID,AD_ViewComponent_UU,Created,CreatedBy,EntityType,IsActive,Name,Updated,UpdatedBy,AD_Table_ID,SeqNo,WhereClause,FromClause,IsUnionAll) VALUES (0,0,200220,'3a3a0b11-e939-4611-b1e8-fc9d6bdd0f04',TO_TIMESTAMP('2022-07-18 15:53:31','YYYY-MM-DD HH24:MI:SS'),100,'D','Y','c_invoice_candidate_v-020',TO_TIMESTAMP('2022-07-18 15:53:31','YYYY-MM-DD HH24:MI:SS'),100,200292,20,'WHERE  rma.docstatus = ''CO''::bpchar
```
Comment (CarlosRuiz-globalqss) [inline migration/iD10/oracle/202207181552_IDEMPIERE-1965.sql:12]: do not use adempiere. schema neither :: casting in the view definition

File: migration/iD10/oracle/202207181552_IDEMPIERE-1965.sql [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+;
+
+-- Jul 18, 2022, 3:57:06 PM CEST
+INSERT INTO AD_ViewColumn (AD_Client_ID,AD_Org_ID,AD_ViewColumn_ID,AD_ViewColumn_UU,Created,CreatedBy,EntityType,IsActive,Updated,UpdatedBy,AD_ViewComponent_ID,ColumnName,ColumnSQL,SeqNo) VALUES (0,0,217395,'a353b097-b281-42d5-85f4-ed6c5e346153',TO_TIMESTAMP('2022-07-18 15:57:06','YYYY-MM-DD HH24:MI:SS'),100,'D','Y',TO_TIMESTAMP('2022-07-18 15:57:06','YYYY-MM-DD HH24:MI:SS'),100,200220,'TotalLines','rma.amt',120)
+;
+
+-- Jul 18, 2022, 4:05:14 PM CEST
+INSERT INTO AD_ViewColumn (AD_Client_ID,AD_Org_ID,AD_ViewColumn_ID,AD_ViewColumn_UU,Created,CreatedBy,EntityType,IsActive,Updated,UpdatedBy,AD_ViewComponent_ID,ColumnName,ColumnSQL,SeqNo) VALUES (0,0,217396,'b34664fd-2289-4224-8677-81f19015e40c',TO_TIMESTAMP('2022-07-18 16:05:13','YYYY-MM-DD HH24:MI:SS'),100,'D','Y',TO_TIMESTAMP('2022-07-18 16:05:13','YYYY-MM-DD HH24:MI:SS'),100,200220,'DocSource','''R''::text',130)
```
Comment (CarlosRuiz-globalqss) [inline migration/iD10/oracle/202207181552_IDEMPIERE-1965.sql:90]: do not use :: casting in view columns

**Comment by PeterTakacs300** [comment]
The last, deprecation commit contains also removals of DocSource element and Reference to avoid collision with the merged https://idempiere.atlassian.net/browse/IDEMPIERE-1963 issue

File: org.adempiere.base.process/src/org/compiere/process/InvoiceGenerate.java [hunk 3]
```diff
@@ -149,8 +149,9 @@ protected String doIt () throws Exception
 			+ ", Consolidate=" + p_ConsolidateDocument);
 		//
 		StringBuilder sql = null;
-		if (p_Selection)	//	VInvoiceGen
+		if (getProcessInfo().getAD_InfoWindow_ID() > 0)	//	VInvoiceGen
```
Comment (hengsin) [inline org.adempiere.base.process/src/org/compiere/process/InvoiceGenerate.java:152]: There are plenty of existing projects that's using a custom "Invoice Generate (Manual)" form and might not want to migrate to the new info window approach immediately. To preserve backward compatibility, it is better to change this to `if ((getProcessInfo().getAD_InfoWindow_ID() > 0) || (getProcessInfo().getAD_InfoWindow_ID()==0 && p_Selection))`.

Also, please remove the "// VInvoiceGen" comment, that's no longer relevant.

File: org.adempiere.base.process/src/org/adempiere/process/InvoiceGenerateRMA.java [hunk 4]
```diff
@@ -74,11 +74,14 @@ else if (name.equals("DocAction"))
                 log.log(Level.SEVERE, "Unknown Parameter: " + name);
         }
         
-        m_dateinvoiced = Env.getContextAsDate(getCtx(), Env.DATE);
-        if (m_dateinvoiced == null)
-        {
-            m_dateinvoiced = new Timestamp(System.currentTimeMillis());
+        if (m_dateinvoiced == null) {
```
Comment (hengsin) [inline org.adempiere.base.process/src/org/adempiere/process/InvoiceGenerateRMA.java:77]: this change is strange, I don't see there's other changes that can make m_dateinvoiced != null at this point.

File: org.adempiere.base.process/src/org/adempiere/process/InvoiceGenerateRMA.java [hunk 5]
```diff
@@ -74,11 +74,14 @@ else if (name.equals("DocAction"))
...
-        {
-            m_dateinvoiced = new Timestamp(System.currentTimeMillis());
+        if (m_dateinvoiced == null) {
+        	m_dateinvoiced = Env.getContextAsDate(getCtx(), Env.DATE);
+	        if (m_dateinvoiced == null)
+	        {
+	        	m_dateinvoiced = new Timestamp(System.currentTimeMillis());
+	        }
         }
+        p_Selection = getProcessInfo().getAD_InfoWindow_ID() > 0;
```
Comment (hengsin) [inline org.adempiere.base.process/src/org/adempiere/process/InvoiceGenerateRMA.java:84]: to keep backward compatiblity, replace this with `if (getProcessInfo().getAD_InfoWindow_ID() > 0) p_Selection=true`

**Comment by norbertbede** [comment]
does any action required on our side ? norbert

**Comment by hengsin** [comment]
[pr_1400.patch.txt](https://github.com/idempiere/idempiere/files/9215244/pr_1400.patch.txt)

- Add unit test for InvoiceGenerate and InvoiceGenerateRMA

**Comment by hengsin** [comment]
[1400.patch.txt](https://github.com/idempiere/idempiere/files/9215304/1400.patch.txt)

- Add Unit Test for InvoiceGenerate and InvoiceGenerateRMA

**Comment by hengsin** [comment]
[1400_migration.patch.txt](https://github.com/idempiere/idempiere/files/9215731/1400_migration.patch.txt)

Fix migration script:
- Wrong GROUP BY clause
- Use IN instead of = ANY (array[...]). IN works for both Oracle and PostgreSQL while = ANY (array[..]) is PostgreSQL only.

**Review by hengsin — APPROVED** [review]
tested

**Comment by CarlosRuiz-globalqss** [comment]
Hi @PeterTakacs300 - can you please add the following patch:
[1965_PR.txt](https://github.com/idempiere/idempiere/files/9233768/1965_PR.txt)

Regards,

Carlos Ruiz

**Review by CarlosRuiz-globalqss — APPROVED** [review]

**Pull request merged**